### ISSUE TITLE
MSSQL unique creates index instead of column constraint

### ISSUE THREAD

**Issue by edwardwu-servian** [knex/knex#4825]
# Environment

Knex version: ^0.95.14
Database + version: Microsoft SQL Azure (RTM) - 12.0.2000.8
OS: macOS 11.6.1

@smorey2 

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do
```javascript
exports.up = function (knex, Promise) {
  return knex.schema
    .withSchema("meta")
    .createTable("constraint_artefact_type", function (table) {
      table.increments("constraint_artefact_type_id");
      table.string("value").notNullable();
      table.unique("value");
    });
};
```
https://github.com/knex/knex/blob/27ade6f8814a5601916489a96474b55649e15462/lib/dialects/mssql/schema/mssql-tablecompiler.js#L307
According to this code, this will generate 
```SQL
CREATE TABLE [meta].[constraint_artefact_type] ([constraint_artefact_type_id] int identity(1,1) not null primary key, [value] nvarchar(255) not null;
CREATE UNIQUE INDEX [constraint_variable_object_type_value_unique] ON [meta].[constraint_variable_object_type] ([value]) WHERE [value] IS NOT NULL;
```
However, this filtered unique index has made it impossible to use `[value]` column in subsequent Foreign Key constraints. Below statement will fail creating foreign key. This is highlighted in current implementation (https://stackoverflow.com/questions/767657/how-do-i-create-a-unique-constraint-that-also-allows-nulls/767702#767702)
```javascript
exports.up = function (knex, Promise) {
  return knex.schema
    .withSchema("meta")
    .createTable("artefact", function (table) {
      table
        .string("type")
        .notNullable();
      table
        .foreign("type")
        .references("value")
        .inTable("meta.constraint_artefact_type")
        .onDelete("SET NULL");
      table.unique(["name", "source"]);
    });
};
```

Expected behaviour here is ` table.unique("value")` would be generating a UNIQUE constraint on `[value]` column (which would be similar to postgres)
```SQL
CREATE TABLE [meta].[constraint_artefact_type] ([constraint_artefact_type_id] int identity(1,1) not null primary key, [value] nvarchar(255) not null UNIQUE;
```
or
```SQL
CREATE TABLE [meta].[constraint_artefact_type] ([constraint_artefact_type_id] int identity(1,1) not null primary key, [value] nvarchar(255) not null;
ALTER TABLE  [meta].[constraint_artefact_type] ADD CONSTRAINT constraint_variable_object_type_value_unique UNIQUE(value);
```
Using UNIQUE constraint, subsequent foreign key constraints would not fail.

2. Error message
```
RequestError: CREATE TABLE [meta].[artefact] (... [type] nvarchar(255) not null... CONSTRAINT [artefact_type_foreign] FOREIGN KEY ([type]) REFERENCES [meta].[constraint_artefact_type] ([value]) ON DELETE SET NULL) - Could not create constraint or index. 
```
MSSQL returns following error if SQL statements provided in debug mode are manually executed:
```
Failed to execute query. Error: There are no primary or candidate keys in the referenced table 'meta.constraint_artefact_type' that match the referencing column list in the foreign key 'artefact_type_foreign'.
```
3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.
```javascript
exports.up = function (knex, Promise) {
  return knex.schema
    .withSchema("meta")
    .createTable("constraint_artefact_type", function (table) {
      table.increments("constraint_artefact_type_id");
      table.string("value").notNullable();
      table.unique("value");
    });
};
```
```javascript
exports.up = function (knex, Promise) {
  return knex.schema
    .withSchema("meta")
    .createTable("artefact", function (table) {
      table
        .string("type")
        .notNullable();
      table
        .foreign("type")
        .references("value")
        .inTable("meta.constraint_artefact_type")
        .onDelete("SET NULL");
      table.unique(["name", "source"]);
    });
};
```

**Comment by jeremy-w** [comment]
I originally thought it peculiar as well, however, on looking more closely, I found that `unique` is documented as creating an index, not a constraint:

https://github.com/knex/documentation/blob/48fcde1d402d13eb8123cefc9ed618e126e74273/sections/schema.js#L822-L826

Where I've wanted a unique constraint, I've used `knex.raw`. However, customizing the details of the index is what the options bucket is there for; the most flexible option is likely allowing to replace the WHERE clause entirely with whatever you want, or set it to some sort of sentinel to remove it.

SQL Server is rather elaborate in its unique index config, while Postgres is quite spartan. Roundup:

- SQL Server: https://docs.microsoft.com/en-us/sql/t-sql/statements/create-index-transact-sql?view=sql-server-ver15
- Oracle: https://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_5010.htm (and function-based unique indexes are kinda neat https://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_5010.htm#BGEHDECJ )
- CockroachDB: https://www.cockroachlabs.com/docs/v21.1/create-index.html
- MariaDB: https://mariadb.com/kb/en/create-index/
- Postgres: https://www.postgresql.org/docs/current/indexes-unique.html

(A "where" clause when creating an index looks to be SQL Server-only to me, aside from CockroachDB inverted index creation.)

**Comment by mlevit** [comment]
@jeremy-w what about a flag on `table.unique` which allows the creation of either an `index` or a `constraint` for MSSQL?

My only concern with `knex.raw` is the interoperability between DB engines.

**Comment by ashley-mallia** [comment]
The current implementation of `table.unique` adding the filter WHERE ([value] IS NOT NULL) in MSSQL allows _any number_ of NULL values to be inserted into that 'unique' column. 
This behaviour sounds like a bug. Why would a column column marked unique not restrict NULL values, just like any other value.
Perhaps provide an option to users for filtering, but by default there should be no filter.

**Comment by jeremy-w** [comment]
It’s not a bug - it’s actually the SQL standard behavior. NULL is not equal to NULL - it’s like asking whether the gifts in two still-wrapped presents are equal or not.

knex uses the implementation it does in order to behave the same as other engines that do follow the standard. It could certainly be unexpected if you are just targeting MSSQL and have not worked with other engines, though.

Would you like to provide a PR that implements the suggested option?
--
Jeremy W. Sherman
http://jeremywsherman.com/

**Comment by kibertoad** [comment]
Released in 1.0.1.
